### PR TITLE
add scANVI example from single cell transcriptomics

### PR DESCRIPTION
**Pull request by martinjankowiak**
references:

[1] "Harmonization and Annotation of Single-cell Transcriptomics data with Deep Generative Models,"
    Chenling Xu, Romain Lopez, Edouard Mehlman, Jeffrey Regier, Michael I. Jordan, Nir Yosef.

[2] https://github.com/YosefLab/scvi-tutorials/blob/50dd3269abfe0c375ec47114f2c20725a016736f/seed_labeling.ipynb

example output:
[scanvi.pdf](https://github.com/pyro-ppl/pyro/files/5287423/scanvi.pdf)

### CODE REVIEW

File: examples/scanvi/data.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    cd4_naive_mask = _get_cell_mask(normalized, cd4_naive_geneset)
+    cd4_mem_mask = _get_cell_mask(normalized, cd4_mem_geneset)
+
+    seed_labels = -np.ones(cd4_mem_mask.shape[0])
+    seed_labels[cd8_naive_mask] = 0  # "CD8 Naive T cell"
+    seed_labels[cd4_naive_mask] = 1  # "CD4 Naive T cell"
+    seed_labels[cd4_mem_mask] = 2    # "CD4 Memory T cell"
+    seed_labels[cd4_reg_mask] = 3    # "CD4 Regulatory T cell"
+
+    # filter out non-variable genes
```
Comment (martinjankowiak) [inline examples/scanvi/data.py:137]: @romain-lopez is this filtering logic reasonable in this context? i'm a big confused about how these scanpy objects work. in particular can i now use `adata` to make UMAP plots using scanpy?

File: examples/scanvi/data.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    adata = adata[:, highly_variable_genes]
+    adata.raw = adata
+
+    Y = torch.from_numpy(seed_labels).long()
+    X = torch.from_numpy(sparse.csr_matrix.todense(adata.X)).float()
+
+    if cuda:
+        X, Y = X.cuda(), Y.cuda()
+
+    # subsample and remove ~90% of the unlabeled cells
```
Comment (martinjankowiak) [inline examples/scanvi/data.py:153]: @romain-lopez i'm also removing some of the unlabeled cells. i'm imagining this is still enough data to learn something reasonable?

File: examples/scanvi/scanvi.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+            z2_loc, z2_scale = self.z2_decoder(z1, y)
+            z2 = pyro.sample("z2", dist.Normal(z2_loc, z2_scale).to_event(1))
+
+            l_scale = self.l_scale * x.new_ones(1)
+            l = pyro.sample("l", dist.LogNormal(self.l_loc, l_scale).to_event(1))
+
+            gate, mu = self.x_decoder(z2)
+            eps = 1.0e-6
+            nb_logits = (theta + eps).log() - (l * mu + eps).log()
```
Comment (martinjankowiak) [inline examples/scanvi/scanvi.py:176]: @romain-lopez i'm still seeing some numerical instabilities but haven't yet investigated in detail where they're coming from. do you recall the major sources of instability in your experiments?

File: examples/scanvi/scanvi.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+        with pyro.plate("batch", len(x)), pyro.poutine.scale(scale=self.scale_factor):
+            z1 = pyro.sample("z1", dist.Normal(0, x.new_ones(self.latent_dim)).to_event(1))
+            y = pyro.sample("y", dist.OneHotCategorical(logits=x.new_zeros(self.num_labels)))
+
+            z2_loc, z2_scale = self.z2_decoder(z1, y)
+            z2 = pyro.sample("z2", dist.Normal(z2_loc, z2_scale).to_event(1))
+
+            l_scale = self.l_scale * x.new_ones(1)
+            l = pyro.sample("l", dist.LogNormal(self.l_loc, l_scale).to_event(1))
```
Comment (martinjankowiak) [inline examples/scanvi/scanvi.py:172]: @romain-lopez could you please point me to the logic you use to set the mean/variance of the log-normal distribution for the  cell-specific scaling factor `l`? i want to make sure this is done correctly

File: examples/scanvi/scanvi.py (hunk 3 above)
Comment (romain-lopez) [reply examples/scanvi/scanvi.py]: (0) The initialization of the params is important. If the variance is too high, you simply can't learn anything
(1) I used some softplus parameterization, not sure ow much that impact things here
(2) we had to do some clamping on the library size (if I remember well)

Let me know if you have residual problems, we can ask some other people as well!

**Comment by romain-lopez** [comment]
Hi Martin, 

That looks cool! Please let me know if that answers all your questions! And don't hesitate to reach out. Let me also know if I can  help with some wording or description if necessary (e.g., if you write some text or describe the model somewhere).

**Review by fritzo — COMMENTED** [review]
- [x] run `make license`
- [x] add spdx licence identifier and copyright line for copied code:
  Copyright+license headers on files adapted from YsefLab should look like this:
  ```py
  # Copyright Contributors to the Pyro project.
  # Copyright (c) 2020, YosefLab.
  # SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
  ```
- [x] add `scanpy` and `scvi` to `EXTRAS_REQUIRE` in setup.py
- [x] add mini dataset for CI

**Comment by martinjankowiak** [comment]
@fritzo how exactly do i "add spdx licence identifier and copyright line for copied code"?

**Comment by martinjankowiak** [comment]
cc @mbabadi

File: examples/scanvi/data.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    adata_filter = adata.copy()
+    sc.pp.normalize_per_cell(adata_filter, counts_per_cell_after=1e4)
+    sc.pp.log1p(adata_filter)
+    sc.pp.highly_variable_genes(adata_filter, min_mean=0.0125, max_mean=3.0, min_disp=0.5)
+    highly_variable_genes = adata_filter.var["highly_variable"]
+
+    Y = torch.from_numpy(seed_labels).long()
+    X = torch.from_numpy(sparse.csr_matrix.todense(adata.X)).float()
+
+    # the prior mean and scale for the log count latent variable `l`
```
Comment (martinjankowiak) [inline examples/scanvi/data.py:186]: @romain-lopez i'm assuming it's as simple as this, right?

**Comment by martinjankowiak** [comment]
@romain-lopez ok i think i'm mostly done with this PR. please have another look if you have a chance! you can see an example of the output in the PR description. looks reasonable to me?

some notes:
- i don't bother to do dropout or KL annealing for simplicity. as far as i can tell this doesn't prevent me from getting reasonable results. (i'm also keeping the neural networks relatively small/simple)
- i'm still reducing the number of genes by a lot but i did indeed find that i need at least ~10k cells to get more reasonable looking results. here i'm using 20k.

**Comment by fritzo** [comment]
@martinjankowiak Copyright+license headers on files adapted from YosefLab should look like this:
```py
# Copyright Contributors to the Pyro project.
# Copyright (c) 2020, YosefLab.
# SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
```

**Comment by martinjankowiak** [comment]
travis is erroring out as follows. i guess `scanpy` depends on `tables` which depends on a  HDF5 installation?

```
Collecting tables
  Downloading tables-3.6.1.tar.gz (4.6 MB)
    ERROR: Command errored out with exit status 1:
     command: /home/travis/virtualenv/python3.5.6/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-i50r7oe8/tables/setup.py'"'"'; __file__='"'"'/tmp/pip-install-i50r7oe8/tables/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-63jjx64u
         cwd: /tmp/pip-install-i50r7oe8/tables/
    Complete output (15 lines):
    /tmp/H5closeqf30ta7r.c:1:1: warning: return type defaults to ‘int’ [-Wimplicit-int]
     main (int argc, char **argv) {
     ^
    /tmp/H5closeqf30ta7r.c: In function ‘main’:
    /tmp/H5closeqf30ta7r.c:2:5: warning: implicit declaration of function ‘H5close’ [-Wimplicit-function-declaration]
         H5close();
         ^
    /usr/bin/ld: cannot find -lhdf5
    collect2: error: ld returned 1 exit status
    * Using Python 3.5.6 (default, Jan  9 2019, 10:45:27)
    * USE_PKGCONFIG: True
    .. ERROR:: Could not find a local HDF5 installation.
       You may need to explicitly state where your local HDF5 headers and
       library can be found by setting the ``HDF5_DIR`` environment
       variable or by using the ``--hdf5`` command-line option.
```

File: examples/scanvi/data.py (hunk 5 above)
Comment (romain-lopez) [reply examples/scanvi/data.py:187]: yes!

**Review by romain-lopez — APPROVED** [review]

**Comment by romain-lopez** [comment]
That looks reasonable to me indeed! Maybe you can add as a disclaimer in the file that this is not meant to exactly reproduce scANVI but to propose a close enough implementation based on pyro, and that can reproduce approximately the result of the figure in the scanvi manuscript :+1:

**Comment by martinjankowiak** [comment]
@fritzo this is ready for review. though will need to figure out a workaround for the scanpy travis issue noted above

**Comment by martinjankowiak** [comment]
i can't reproduce the travis error in torch 1.6.  i guess there's a type conversion difference between 1.5 and 1.6 that's triggering the error?
```
File "/home/travis/build/pyro-ppl/pyro/examples/scanvi/scanvi.py", line 234, in guide
    z1_loc, z1_scale = self.z1_encoder(z2, y)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/torch/nn/modules/module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/travis/build/pyro-ppl/pyro/examples/scanvi/scanvi.py", line 121, in forward
    z2_y = torch.cat(z2_y, dim=-1)
RuntimeError: Expected object of scalar type float but got scalar type long int for sequence element 1.
```

File: examples/scanvi/scanvi.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+    def forward(self, z1, y):
+        z1_y = torch.cat([z1, y], dim=-1)
+        # we reshape the input to be two-dimensional so that nn.BatchNorm1d behaves correctly
+        _z1_y = z1_y.reshape(-1, z1_y.size(-1))
+        hidden = self.fc(_z1_y)
+        # if the input was three-dimensional we now restore the original shape
+        hidden = hidden.reshape(z1_y.shape[:-1] + hidden.shape[-1:])
+        loc, scale = split_in_half(hidden)
+        # here and elsewhere softplus ensures that scale is positive
```
Comment (fritzo) [inline examples/scanvi/scanvi.py:79]: nit: add a comment (and punctuation)?

> note softplus is more numerially stable than exp

File: examples/scanvi/scanvi.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+        scanvi.cuda()
+
+    # setup an optimizer
+    optim = Adam({"lr": args.learning_rate})
+
+    # tell Pyro to enumerate out y when y is unobserved
+    guide = config_enumerate(scanvi.guide, "parallel", expand=True)
+
+    # setup a variational objective for gradient-based learning
+    svi = SVI(scanvi.model, guide, optim, TraceEnum_ELBO())
```
Comment (fritzo) [inline examples/scanvi/scanvi.py:264]: can you mention why we use `TraceEnum_ELBO()` rather than `Trace_ELBO()`

File: examples/scanvi/scanvi.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        fig.colorbar(s21, ax=axes[2, 1])
+
+        fig.tight_layout()
+        plt.savefig('scanvi.pdf')
+
+
+if __name__ == "__main__":
+    assert pyro.__version__.startswith('1.4.0')
+    # parse command line arguments
+    parser = argparse.ArgumentParser(description="parse args")
```
Comment (fritzo) [inline examples/scanvi/scanvi.py:336]:
```py
description="single-cell ANnotation using Variational Inference"
```

**Review by fritzo — COMMENTED** [review]
Looks great, I have only minor comments. Thanks for creating this!!

In a follow-up PR I think it would be good to display this in pyro.ai/examples, ideally with a plot as we did with the [sir_hmc](http://pyro.ai/examples/sir_hmc.html) example.

File: examples/scanvi/scanvi.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+
+from data import get_data
+
+
+# helper for making fully-connected neural networks
+def make_fc(dims):
+    layers = []
+    for in_dim, out_dim in zip(dims, dims[1:]):
+        layers.append(nn.Linear(in_dim, out_dim))
+        layers.append(nn.BatchNorm1d(out_dim))
```
[the remark this answers is not in the document]
Comment (martinjankowiak) [reply examples/scanvi/scanvi.py:47]: yeah i'm not sure that's a difficult question to answer. in practice i think batch norm can be pretty helpful for numerically stabilizing neural networks with e.g. high dimensional outputs

**Review by fritzo — APPROVED** [review]

**Pull request merged**